### PR TITLE
Fix bugs to enable SAGEMAKER_GUNICORN_WORKERS > 1 on MME.

### PR DESCRIPTION
**Pull request by chen3933**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description

Fix bugs for tensorflow inference to enable SAGEMAKER_GUNICORN_WORKERS > 1 on MME

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [x] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

### Test

[827e33f](https://github.com/aws/deep-learning-containers/pull/2827/commits/827e33feb2454774d87f2438b7110e09ba722b68) tf 2.11

- Sanity Test
- EC2 test
- ECS test
- EKS test
- Sagemaker test
- Sagemaker Local test

[249ab89](https://github.com/aws/deep-learning-containers/pull/2827/commits/249ab8938cf1b6d85f09ef13da0b52696c6dd580) tf 2.10

- All Above

[8cc2f96](https://github.com/aws/deep-learning-containers/pull/2827/commits/8cc2f968c7f25c6bdc72f49f09f876be54236a71) graviton 2.9

- All Above

[0907e3a](https://github.com/aws/deep-learning-containers/pull/2827/commits/0907e3a2502f1d5631c6d61dd922c51de7ee9ca8) tf 2.9

- All Above

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

### CODE REVIEW

File: tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py:286]: with timeout(timeout_seconds):
        while True:
            try:
                session = requests.Session()
                retries = Retry(total=9, backoff_factor=0.1)
                session.mount("http://", requests.adapters.HTTPAdapter(max_retries=retries))
                response = session.get(tfs_url)
                ...
            except (
                ...

The above is the original code. During the `retry`, if `with timeout` throw exception, it will be caught by `retry` and increase retry counter by 1. In other words, if tfs fails to start, this function will retry indefinitely. 

Since requests.Session() does not provide a total timeout. the walk around is base on the total timeout to calculate retry count and limit each request with timeout 100ms.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:320]: Updated PR to use `lock = manager.Lock()` instead.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:505]: Updated PR to use 
```
from multiprocessing import Manager
manager = Manager()
```
instead.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:116]: When customer sends invoke endpoint request. Sagemaker will make following calls to TFS DLC:
1. Load model request. ( this request will load `SAGEMAKER_TFS_INSTANCE_COUNT` number of tfs instance )
2. Invoke endpoint.
The total timeout is 60s.

Since we are starting tfs instance one by one, we give each tfs 55/tfs_count seconds to start. the rest 5s to run inference.

I am not sure why it was 300s in the first place. By given the original code `wait_for_model` (https://github.com/aws/deep-learning-containers/blob/master/tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py#L257) is honor timeout. I am not surprised if they just put a random number there.

File: tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py:270]: This timeout is on each request level. We are going to retry multiple times to verify if tfs started.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (davidthomas426) [inline tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:326]: I think some of your locking is pretty course-grained, particularly around the whole loadmodel. Be aware that if an invocation comes in for a non-existent model during a load_model request, it will be blocked on acquiring the lock until the model is finished loading.

That's probably enough of a corner case that it's fine, though, I guess. I just wanted to point that out about the behavior for your awareness, at least.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (davidthomas426) [inline tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:504]: I'm not convinced that we need the pickling here at all. I think we could instead just put `value` directly in the manager dict and it would work.

File: tensorflow/inference/docker/build_artifacts/sagemaker/serve.py
Comment (davidthomas426) [inline tensorflow/inference/docker/build_artifacts/sagemaker/serve.py:222]: One question I have about gunicorn CLI vs doing it with `.run()` python API: is there any different in the ability of gunicorn to restart workers automatically when they die? Just check on that before merge...

File: tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py
Comment (davidthomas426) [reply tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py:270]: Yes, but any model that takes longer than 100ms to load will never complete the request, with this code. Right?

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (davidthomas426) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:116]: I don't love this, as it feels unnecessarily restrictive. But I guess maybe it's fine, as long as `tfs_count` doesn't get too high so that the individual timeouts are unreasonably small.

File: tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py
Comment (davidthomas426) [reply tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py:286]: Yes, the `while True` loop is broken, I agree. I get what you are trying to calculate here. I'm saying that the calculation isn't right anyway, because it only counts the time between requests, not including the actual request time itself. So your calculated retry count is going to be too high if the model load requests take a significant amount of time, making it kind of pointless.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (davidthomas426) [inline tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:212]: Is this requirement documented somewhere, where "70" is considered the right value?

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:326]: Yes, Thank you for point it out.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:212]: I did some experiment and try to load multiple large models to trigger OOM issue. The lowest memory threshold which triggers OOM is 73% of CPU memory usage. In normal use case, sagemaker will monitor memory usage and issue a load model request.

File: tensorflow/inference/docker/build_artifacts/sagemaker/serve.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/serve.py:226]: Hmm, I didn't see any mention or warning about difference of using cli vs python API in their doc : https://docs.gunicorn.org/en/stable/custom.html#direct-usage-of-existing-wsgi-apps

File: tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py:276]: For each attempt, we timeout in 100 ms. And we are using retry to wait model to load.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:116]: Customer can overwrite timeout with `SAGEMAKER_TFS_WAIT_TIME_SECONDS `. But given sagemaker has hard limit (60s) on each request. I think it will timeout if customer is trying to start too many tfs instances on sagemaker endpoint.

File: tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py:319]: Sync up offline. The following is the problem statement:

In order to use `multiprocess.manager.dict` to we have to use gthread instead of gevent. 
Here is the issue using gevent with `multiprocess.manager.dict` : https://github.com/benoitc/gunicorn/issues/1407

However, timeout using signal not work with `gthread` with following error message:
```
2023-04-04 16:05:22 [FALCON] [ERROR] GET /increment/abc => Traceback (most recent call last):
  File "falcon/app.py", line 365, in falcon.app.App.__call__
  File "/sagemaker_tmp/test.py", line 53, in on_get
    with timeout(2):
  File "/usr/local/lib/python3.9/contextlib.py", line 119, in __enter__
    return next(self.gen)
  File "/sagemaker_tmp/test.py", line 43, in timeout
    signal.signal(signal.SIGALRM, _raise_timeout_error)
  File "/usr/local/lib/python3.9/signal.py", line 56, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
ValueError: signal only works in main thread of the main interpreter
```

For now, we are calculating timeout. Added comment about small delta we might have.

File: tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py
Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/tfs_utils.py]: Fixed. Thank you for point it out.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (davidthomas426) [inline tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:54]: Don't we need to avoid creating a `multiprocessing.Manager()` if we're using `gevent`?

Probably good to add a quick validation here in this file, too, that ensures we don't somehow end up with `gevent` in MME mode.

Comment (chen3933) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:54]: As long as we don't call it. It will be fine.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (davidthomas426) [reply tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:322]: ok, this could technically still be a degredation in MME with instance count 1, and it's unnecessary in that case since there's nothing to sync with.

Could just add an "if instance count is 1" check in that if statement. Alternatively, you could modify "lock" to be a no-op unless in MME insance_count > 1 case.

**Review by davidthomas426 — CHANGES_REQUESTED** [review]
Avoid adding locking for use cases that worked before and that don't need locks.

File: dlc_developer_config.toml
Comment (chen3933) [reply dlc_developer_config.toml]: We need to turn it on manually and enable test to run in separate commit. Before merge the change, we need to turn them back off. Otherwise `dlc-pr-quick-checks` will fail. 

I have several commit to cover different framework version (2.9, 2.10, 2.11) as well as different platform (x86, graviton). In the last commit. turn off all the test for merge.

File: dlc_developer_config.toml
Comment (chen3933) [reply dlc_developer_config.toml]: We need to turn it on manually and enable test to run in separate commit. Before merge the change, we need to turn them back off. Otherwise dlc-pr-quick-checks will fail.

I have several commit to cover different framework version (2.9, 2.10, 2.11) as well as different platform (x86, graviton). In the last commit. turn off all the test for merge.

File: dlc_developer_config.toml
Comment (chen3933) [reply dlc_developer_config.toml]: Same as above.

File: tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py
Comment (davidthomas426) [inline tensorflow/inference/docker/build_artifacts/sagemaker/python_service.py:579]: Oh, looks like you did already do this, check `args.workers` on line 566, coming from `serve.py` line 226

**Review by davidthomas426 — APPROVED** [review]
LGTM

**Review by xncqr — APPROVED** [review]
Tests look good. LGTM

**Pull request merged**
